src/sshagent/SSHAgent.cpp

```diff
@@ -439,11 +437,13 @@ void SSHAgent::databaseLocked(QSharedPointer<Database> db)
         }
         OpenSSHKey key = it.key();
         if (it.value().second) {
-            if (!removeIdentity(key)) {
+            if (removeIdentity(key)) {
+                m_addedKeys.erase(it);
+            } else {
                 emit error(m_error);
             }
         }
-        it = m_addedKeys.erase(it);
```

Comment:
The reason why we unconditionally forget about all keys on lock is that we shouldn't keep them in-memory in locked state even though they should only contain the public part as technically we shouldn't know what keys each database contain in that state.

If a key is already in agent when you unlock the database the next time and you didn't have "remove on lock" bit set then it's expected adding a key will fail.

However we could either buffer the errors instead of emitting them immediately so it doesn't stop dead or we ignore errors on lock completely.